### PR TITLE
Add `merge` materialization

### PR DESCRIPTION
**Pull request by hovaesco**
## Overview
Add `merge` materialization

## Checklist

- [ ] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change
- [x] I have run `changie new` to [create a changelog entry](https://github.com/starburstdata/dbt-trino/blob/master/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -209,6 +209,24 @@ select * from {{ ref('users') }}
 {% endif %}
 ```
 
+##### `merge`
+
+Through the `merge` incremental strategy, dbt-trino constructs `MERGE` statement which `INSERT` new records and `UPDATE` existing ones based on unique key specified by `unique_key`.
```
Comment (mdesmet) [inline README.md:214]: Also add a note that they can use the `delete+insert` materialization if their `unique_key` is not unique

**Review by mdesmet — COMMENTED** [review]
Also merge tests for Starburst should be disabled

File: README.md (hunk 1 above)
Comment (findinpath) [inline README.md:214]: add a link for `MERGE` https://trino.io/docs/current/sql/merge.html

Do emphasize the fact that `MERGE` is not available on all the connectors.

File: README.md [hunk 2]
```diff
@@ -209,6 +209,24 @@ select * from {{ ref('users') }}
...
+
+```jinja2
+{{
+    config(
+      materialized = 'incremental',
+      unique_key='user_id',
+      incremental_strategy='merge',
+      )
+}}
+select * from {{ ref('users') }}
```
Comment (findinpath) [inline README.md:224]: Please do add also an example with `incremental_predicates`

File: README.md (hunk 1 above)
Comment (findinpath) [inline README.md:214]: Do note that `MERGE` does support also `DELETE` , not only `UPDATE` for dealing with existing entries.

File: tests/functional/adapter/materialization/test_incremental_merge.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    @pytest.fixture(scope="class")
+    def project_config_update(self):
+        return {
+            "name": "incremental",
+            "models": {"+incremental_strategy": "merge"},
+            "seeds": {"incremental": {"seed": {"+column_types": {"some_date": "date"}}}},
+        }
+
+    # Below tests are disabled due to bug in Trino MERGE statement when merge_condition is False
+    # TODO https://github.com/trinodb/trino/issues/13751
```
Comment (findinpath) [inline tests/functional/adapter/materialization/test_incremental_merge.py:137]: Cool. You are improving `MERGE` on Trino ❤️

File: tests/functional/adapter/materialization/test_incremental_merge.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    @pytest.fixture(scope="class")
+    def project_config_update(self):
+        return {
+            "name": "incremental",
+            "models": {"+on_table_exists": "drop", "+incremental_strategy": "merge"},
+            "seeds": {"incremental": {"seed": {"+column_types": {"some_date": "date"}}}},
+        }
+
+    # Below tests are disabled due to bug in Trino MERGE statement when merge_condition is False
+    def test__no_unique_keys(self, project):
```
[the remark this answers is not in the document]
Comment (hovaesco) [reply tests/functional/adapter/materialization/test_incremental_merge.py:163]: Done

File: README.md (hunk 2 above)
Comment (hovaesco) [reply README.md:226]: I disabled it from macro since it's a part of the dbt refactor which is not yet out.

File: README.md [hunk 5]
```diff
@@ -209,6 +209,26 @@ select * from {{ ref('users') }}
 {% endif %}
 ```
 
+##### `merge`
+
+Through the `merge` incremental strategy, dbt-trino constructs [`MERGE` statement](https://trino.io/docs/current/sql/merge.html) which `INSERT` new records and `UPDATE` existing ones based on unique key specified by `unique_key`.
```
Comment (mdesmet) [inline README.md:214]:
```suggestion
Through the `merge` incremental strategy, dbt-trino constructs a [`MERGE` statement](https://trino.io/docs/current/sql/merge.html) which `INSERT`s new and `UPDATE`s existing records based on the unique key specified by `unique_key`. 
```

**Review by mdesmet — APPROVED** [review]
LGTM % docs nit

File: README.md [hunk 6]
```diff
@@ -25,7 +25,7 @@ This repository represents a fork of the [dbt-presto](https://github.com/dbt-lab
 
 ### Compatibility
 
-This dbt plugin has been tested against `Trino` version `386` and `Starburst Enterprise` version `385-e`.
```
Comment (findinpath) [inline README.md:28]: As a follow-up we should test `merge` also against `393-e` release

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 7]
```diff
@@ -59,6 +61,54 @@
...
+                {% do predicates.append(this_key_match) %}
+            {% endfor %}
+        {% else %}
+            {% set unique_key_match %}
+                DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
+            {% endset %}
+            {% do predicates.append(unique_key_match) %}
+        {% endif %}
+    {% else %}
+        {% do predicates.append('FALSE') %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:86]: https://github.com/trinodb/trino/issues/13751#issue-1344466027

Shouldn't we do the `INSERT` instead of the `MERGE` here in `dbt-trino`?

Why should we make use of an edge of `MERGE` when we can very well benefit of `INSERT`?

**Review by findinpath — CHANGES_REQUESTED** [review]
Please convert the use case when there are no unique keys specified to an `INSERT`

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 8]
```diff
@@ -59,6 +61,54 @@
...
+    {% endif %}
+
+    {{ sql_header if sql_header is not none }}
+
+    merge into {{ target }} as DBT_INTERNAL_DEST
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:91]:
```
merge into "iceberg"."test16611829018484321781_test_incremental_merge"."trinary_unique_key_list" as DBT_INTERNAL_DEST
        using "iceberg"."test16611829018484321781_test_incremental_merge"."trinary_unique_key_list__dbt_tmp" as DBT_INTERNAL_SOURCE
        on 
                    DBT_INTERNAL_SOURCE.state = DBT_INTERNAL_DEST.state
                 and 
                    DBT_INTERNAL_SOURCE.county = DBT_INTERNAL_DEST.county
                 and 
                    DBT_INTERNAL_SOURCE.city = DBT_INTERNAL_DEST.city
                

    
    when matched then update set
        "state" = DBT_INTERNAL_SOURCE."state","county" = DBT_INTERNAL_SOURCE."county","city" = DBT_INTERNAL_SOURCE."city","last_visit_date" = DBT_INTERNAL_SOURCE."last_visit_date"
    

    when not matched then insert
        ("state", "county", "city", "last_visit_date")
    values
        (DBT_INTERNAL_SOURCE."state", DBT_INTERNAL_SOURCE."county", DBT_INTERNAL_SOURCE."city", DBT_INTERNAL_SOURCE."last_visit_date")
```

Cosmetically this output - for `test__trinary_unique_key_list` test - (even though purely generated) can be formatted to make it a bit easier to read by humans.

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 9]
```diff
@@ -59,6 +61,54 @@
...
+        using {{ source }} as DBT_INTERNAL_SOURCE
+        on {{ predicates | join(' and ') }}
+
+    {% if unique_key %}
+    when matched then update set
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:96]: Why should we do `set` the unique key components when we know that they are equal with the source row?

Interestingly I see the same behavior in dbt-core https://github.com/dbt-labs/dbt-core/blob/1071a4681df91633301fdf23e34de819b66fbeb7/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L36-L39 for `merge`. 
Am I missing something ?

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 10]
```diff
@@ -69,6 +119,7 @@
 
   {#-- Validate early so we don't run SQL if the strategy is invalid --#}
   {%- set strategy = validate_get_incremental_strategy(config) -%}
+  {% set incremental_predicates = config.get('incremental_predicates', none) %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/incremental.sql:122]: @hovaesco  if we don't use the `incremental_predicates` there's no reason to read them. this is probably a left-over.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 9 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql:96]: Do we expect any performance improvement by omitting the unique key components?

Comment (findinpath) [reply dbt/include/trino/macros/materializations/incremental.sql:96]: Being fairly inexperienced myself in this regard, I asked for feedback on Trino's slack 

https://trinodb.slack.com/archives/CP1MUNEUX/p1661228413753349

tldr; 

>  The columns you don’t update, are still passed thru (identity), so it’s the same situation.

@hovaesco please consider my request optional. In the current situation i find it rather cosmetical - an improvement for the human readers :)

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 7 above)
Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql]: Done, now all tests are enabled by default.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 8 above)
Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql]: Formatting changed to:
```
merge into "iceberg"."test16612448099870284150_test_incremental_merge"."trinary_unique_key_list" as DBT_INTERNAL_DEST 
using "iceberg"."test16612448099870284150_test_incremental_merge"."trinary_unique_key_list__dbt_tmp" as DBT_INTERNAL_SOURCE on
        DBT_INTERNAL_SOURCE.state = DBT_INTERNAL_DEST.state and
        DBT_INTERNAL_SOURCE.county = DBT_INTERNAL_DEST.county and
        DBT_INTERNAL_SOURCE.city = DBT_INTERNAL_DEST.city

        when matched then update set
            "state" = DBT_INTERNAL_SOURCE."state",
            "county" = DBT_INTERNAL_SOURCE."county",
            "city" = DBT_INTERNAL_SOURCE."city",
            "last_visit_date" = DBT_INTERNAL_SOURCE."last_visit_date"

        when not matched then insert
            ("state", "county", "city", "last_visit_date")
        values
            (DBT_INTERNAL_SOURCE."state", DBT_INTERNAL_SOURCE."county", DBT_INTERNAL_SOURCE."city", DBT_INTERNAL_SOURCE."last_visit_date")

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 9 above)
Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql]: Thanks for raising the topic with the Trino community. This is basically a copy and paste from dbt-core.

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 10 above)
Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql]: Done

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 7 above)
Comment (mdesmet) [reply dbt/include/trino/macros/materializations/incremental.sql]: Great catch @findinpath

**Review by findinpath — APPROVED** [review]
